File: python/src/fds/protobuf/stach/extensions/v2/ColumnOrganizedStachExtension.py
import pandas as pd
from fds.protobuf.stach.v2.Package_pb2 import Package

from fds.protobuf.stach.extensions.IStachExtension import IStachExtension
from fds.protobuf.stach.extensions.v2.StachUtilities import StachUtilities


class ColumnOrganizedStachExtension(IStachExtension):

    def __init__(self, package: Package):
        self.package = package

    def get_package(self):
        return self.package
    
    def convert_to_dataframe(self):
        tables = list()
        for primary_table_id in self.package.primary_table_ids:
            tables.append(self.__generate_table(self.package, primary_table_id))
        return tables

    def get_metadata(self):
        metadata = list()
        for primary_table_id in self.package.primary_table_ids:
            metadata.append(self.__generate_metadata(self.package, primary_table_id))
        return metadata

    @staticmethod
    def __generate_table(package_response, primary_table_id):
        """
        The purpose of this function is to generate Tables for the provided stach data through the package.

        :param package_response: Stach Data which is represented as a Package object.
        :param primary_table_id: Refers to the id for a particular table inside a package.
        :return: Returns the generated Table from the package provided.
        """
        if isinstance(package_response, Package):
            primary_table = package_response.tables[primary_table_id]
            header_id = primary_table.definition.header_table_id
            headers = list(list())
            header_table = package_response.tables[header_id]

            if (len(header_id) > 0):

                # Construct header rows
                for header_table_definition in header_table.definition.columns:
                    header_row = list()

                    for primary_table_definition in primary_table.definition.columns:
                        if (primary_table_definition.is_dimension == True):
                            header_row.append(primary_table_definition.description or primary_table_definition.name)
                            continue

                        headerColumnId = primary_table_definition.header_id
                        indexOfHeaderColumnId = [index for index in range(len(header_table.data.rows)) if
                                                header_table.data.rows[index].id == headerColumnId][0]
                        val = header_table.data.columns[header_table_definition.id].values[indexOfHeaderColumnId]
                        header_row.append(val)

                    headers.append(header_row)
            
            else:
                # if there is no headers table process headers from description section
                header = list()
                for column_definition in primary_table.definition.columns:
                    header.append(column_definition.description or column_definition.name)
                headers.append(header)

            # Construct data rows
            data = list(list())
            rowCount = len(primary_table.data.rows)

            # Handling when the stach output doesnt have primarytable.data.rows
            if(rowCount == 0):
                iterator = iter(primary_table.data.columns)
                first_key = next(iterator)
                first_value = primary_table.data.columns.get(first_key)
                rowCount = len(first_value.values.values)

            for i in range(0, rowCount, 1):
                data_row = list()
                for primary_table_definition in primary_table.definition.columns:
                    val = primary_table.data.columns[primary_table_definition.id].values[i]
                    val = StachUtilities.get_value(val)
                    data_row.append(val if val is not None else primary_table_definition.format.null_format)
                data.append(data_row)

            if len(header_table.definition.columns) > 1:
                data_frame = pd.DataFrame(data=data)
                data_frame.columns = pd.MultiIndex.from_arrays(headers)
            else:
                data_frame = pd.DataFrame(data=data, columns=headers[0])
            return data_frame

    @staticmethod
    def __generate_metadata(package_response, primary_table_id):
        """
        The purpose of this function is to generate the metadata for the provided stach data through the package.

        :param package_response: Stach Data which is represented as a Package object.
        :param primary_table_id: Refers to the id for a particular table inside a package.
        :return: Returns the generated metadata from the package provided.
        """
        primary_table = package_response.tables[primary_table_id]
        metadata = {}

        for location in primary_table.data.metadata.locations.table:
            if (len(primary_table.data.metadata.items[location].value.list_value.values) > 0):
                values = list()
                for value in primary_table.data.metadata.items[location].value.list_value.values:
                    values.append(value)
                metadata[location] = values
        return metadata

File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/V2/ColumnOrganizedStachExtension.cs
using System.Collections.Generic;
using System.Linq;
using FactSet.Protobuf.Stach.Extensions.Models;
using System;
using System.Diagnostics;
using Google.Protobuf.WellKnownTypes;

namespace FactSet.Protobuf.Stach.Extensions.V2
{
    public class ColumnOrganizedStachExtension : IStachExtension
    {
        private readonly Stach.V2.Package pkg;
        
        public ColumnOrganizedStachExtension(Stach.V2.Package pkg)
        {
            this.pkg = pkg;
        }
        
        public List<Models.Table> ConvertToTable()
        {
            var tables = new List<Models.Table>();
            foreach (var primaryTableId in pkg.PrimaryTableIds)
            {
                tables.Add(GenerateTable(pkg, primaryTableId));
            }
            return tables;
        }
        
        /// <summary>
        /// The purpose of this function is to generate Table for a given table id in the provided
        /// stach data through the package.
        /// </summary>
        /// <param name="package"></param>
        /// <param name="primaryTableId"></param>
        /// <returns>Returns the generated Table from the package provided.</returns>
        private static Models.Table GenerateTable(Stach.V2.Package package, string primaryTableId)
        {
            
            var primaryTable = package.Tables[primaryTableId];
            var headerId = primaryTable.Definition.HeaderTableId;
            var columnIds = primaryTable.Definition.Columns.Select(c => c.Id).ToList();
            var rowCount = primaryTable.Data.Rows.Count;

            if (rowCount == 0)
            {
                rowCount = primaryTable.Data.Columns.FirstOrDefault().Value.Values.Values.Count;
            }

            var table = new Models.Table
            {
                Rows = new List<Row>(),
                Metadata = new Dictionary<string, string>(),
                RawMetadata = new Dictionary<string, List<Value>>()
            };

            if (headerId.Length > 0)
            {
                var headerTable = package.Tables[headerId];

                var headerDataRowList = headerTable.Data.Rows.Select(x => x.Id).ToList();
                foreach (var headerTableSeriesDefinition in headerTable.Definition.Columns)
                {
                    var headerRow = new Row { Cells = new List<string>() };
                    foreach (var primaryTableSeriesDefinition in primaryTable.Definition.Columns)
                    {
                        if (primaryTableSeriesDefinition.IsDimension)
                        {
                            var description = string.IsNullOrEmpty(primaryTableSeriesDefinition.Description)
                                ? primaryTableSeriesDefinition.Name
                                : primaryTableSeriesDefinition.Description;

                            headerRow.Cells.Add(description);
                            continue;
                        }

                        var index = headerDataRowList.IndexOf(primaryTableSeriesDefinition.HeaderId);
                        var value = headerTable.Data.Columns[headerTableSeriesDefinition.Id].Values.Values[index];

                        headerRow.Cells.Add(StachUtilities.ValueToString(value));
                    }

                    headerRow.isHeader = true;
                    table.Rows.Add(headerRow);
                }
            }
            else
            {
                // if there is no headers table process headers from description section
                var headerRow = new Row { Cells = new List<string>() };
                foreach (var columnDefinition in primaryTable.Definition.Columns)
                {
                    var description = string.IsNullOrWhiteSpace(columnDefinition.Description)
                        ? columnDefinition.Name
                        : columnDefinition.Description;
                    headerRow.Cells.Add(description);
                }

                headerRow.isHeader = true;
                table.Rows.Add(headerRow);
            }


            // Constructs the column data
            for (int i = 0; i < rowCount; i++)
            {
                var dataRow = new Row {Cells = new List<string>()};
                foreach (var columnId in columnIds)
                {
                    dataRow.Cells.Add(
                        StachUtilities.ValueToString(primaryTable.Data.Columns[columnId].Values.Values[i]) ?? string.Empty);
                }

                table.Rows.Add(dataRow);
            }

            if (primaryTable.Data.Metadata != null)
            {
                var metadataItems = primaryTable.Data.Metadata.Items;
                var tableMetadataLocations = primaryTable.Data.Metadata.Locations.Table;
                foreach (var location in tableMetadataLocations)
                {
                    metadataItems.TryGetValue(location, out var metadataItem);
                    if (metadataItem != null)
                    {
                        table.Metadata.Add(location, StachUtilities.ValueToString(metadataItem.Value));

                        List<Value> valuesList = new List<Value>();
                        if (metadataItem.Value.KindCase.ToString() == "ListValue")
                        {
                            foreach (Value val in metadataItem.Value.ListValue.Values)
                            {
                                valuesList.Add(val);
                            }
                        }
                        else if (metadataItem.Value.KindCase.ToString() == "StringValue")
                        {
                            valuesList.Add(metadataItem.Value);
                        }

                        table.RawMetadata.Add(location, valuesList);
                    }
                } 
            }

            return table;
        }

    }
}